### PR TITLE
Switch to tesselated fill rendering with earcut

### LINKED ISSUE #682: tesselate polygons instead of rendering with stencil buffer
We should draw polygons by tesselating them instead of using the stencil buffer trick. Native already does this. https://github.com/mapbox/mapbox-gl-js/commit/7b353122b46503b57e6d5073184883c1c2e21404 is a first attempt that uses libtess.js. Rendering is faster but the tesselation is slow. The buildings layer can take up to 500ms to tesselate.

- take a look at poly2tri.js (which only accepts simple polygons)
- revisit tesselating polygons in the tiles

### LINKED ISSUE #2269: Areas covered by "n % 8 == 0" polygons are not drawn
On 0.15.0, we have some weird artifacts when multiple geometries on the same layer are rendered and happen to overlap each other. 

screen shot 2016-03-14 at 11 04 42 pm

screen shot 2016-03-14 at 11 06 08 pm

You can also view it live at https://app.airmap.io and searching for **Anchorage** with **Recreational - 5 mile radius** airports turned on.

Any ideas?

### PR DESCRIPTION
**Pull request by mourner**
A fresh recut of #948. Had to squish into one commit because all previous branches were a big mess. cc @jfirebaugh @ansis @kkaefer 

fixes #2269
fixes #682

### CODE REVIEW

**Comment by mourner** [comment]
All good except one fill pattern render test! Rendering is super-blazing-fast.

At first I thought some sliver lines I saw were earcut artifacts, but they turn out to be present on master as well:
image

**Comment by mourner** [comment]
The fill-pattern/zoomed test [looks very good](https://circle-artifacts.com/gh/mapbox/mapbox-gl-js/1123/artifacts/0/home/ubuntu/mapbox-gl-js/node_modules/mapbox-gl-test-suite/render-tests/index.html) actually, and there definitely something fishy about the diff algorithm. Should we add it to ignore list @jfirebaugh? Otherwise the PR looks great.

**Comment by jfirebaugh** [comment]
We should update the expected rendering for all of the fill tests.

**Comment by lucaswoj** [comment]
@jfirebaugh @mourner Great work! Do you have an estimate on how long it will be until this can land in a release? It resolves #930, which is blocking a client of ours.

**Comment by jfirebaugh** [comment]
The blocker for adopting earcut is getting confidence that the tessellation is good enough. This is a function of both the quality of our vector tile generation pipeline, and the robustness of earcut itself, and myself, @mourner, @lbud, @flippmoke, and @springmeyer are all pushing on this at the moment. We should have a better estimation next week.

**Comment by lucaswoj** [comment]
Rebased this on master (with #1645) and pushed to `earcut-final-final`.

**Comment by jfirebaugh** [comment]
@lucaswoj Looks like `earcut-final-final` is still pointing at the same commit as `earcut-final`. Feel free to just force push `earcut-final`.

**Comment by lucaswoj** [comment]
Ok. Force pushed here and deleted `earcut-final-final`. (Just in case, previous SHA was 44ce255)

**Comment by mourner** [comment]
@lucaswoj let's rebase on master to make sure it's in mergeable state

**Comment by lucaswoj** [comment]
Done @mourner :rocket:

**Comment by lucaswoj** [comment]
We can close #930 when this lands

**Comment by averas** [comment]
Do you have a fresh estimate on when this could be merged? I'm making heavy use of fill-layers and this branch is much more performant than <code>master</code>.

**Comment by mourner** [comment]
@averas no estimate yet, but we're working on it. It's a challenge since most existing vector tiles are not compatible with the change — it requires strictly simple polygons with a specific winding order (described in a newly released vector tile spec version 2.0).

**Comment by averas** [comment]
@mourner Thanks for the update. Would you say it's safe to keep rebasing this branch on <code>master</code> and using it in production while anticipating it will be merged up ahead? Again, performance is so much better, and personally I haven't seen any glitches with <code>street's</code> tiles.

**Comment by averas** [comment]
Could possibly any of the core developers rebase this branch to master? I gave it a shot but there were some conflicts deep down in the rendering code that I was unable to resolve.

Also, in #2021 @ansis mention that this branch + depth buffer pixel dropping would give me best performance with many polygon fills. It is, however, a bit unclear whether the depth buffering functionality is part of master or another branch. Would this branch rebased to master give me both?

**Comment by mourner** [comment]
Yes, it's @lucaswoj's hobby to rebase this branch once a month — let's leave this to him again. :)

**Comment by lucaswoj** [comment]
Thanks @mourner! Looking forward to doing that rebase sometime today. :smile:

**Comment by lucaswoj** [comment]
Rebase done but there are some failing test-suite tests having to do with un-aliased fill rendering. Investigating now.

**Comment by lucaswoj** [comment]
My impression is that there has always been different antialiasing on this branch vs master but it looks substantially worse with latest rebase. Any idea why this might be @mourner @ansis @jfirebaugh? The previous branch head was da1bd60 (which was before layer-by-layer).

**Comment by averas** [comment]
@lucaswoj Testing this now I also see some anomalies. I also noted, however, that #2060 actually looks better (although not perfect).

**Comment by averas** [comment]
I'm not sure to what extent you are interested in keeping this branch in a mergable state, but I found a more severe anomaly when I placed my (<1 opacity) vt-layers on top of the satellite raster layer. This behaviour is not present in 0.13.1 even though this branch is now rebased against master:

vt-on-raster

**Comment by lucaswoj** [comment]
@averas I am keen on keeping this mergable. I wonder if this problem has the same root cause as the antialiasing problem.

**Comment by averas** [comment]
@lucaswoj Cool.

1) I found that it was not related to the raster layer combo which I first though, it just happened to become much more clear with the contrast it provided. I see the same on top of my regular vector map.

2) I see the same behaviour with a GeoJSON based fill with opacity <1, so it's probably not related to the vector tiles.

**Comment by dchersey** [comment]
I'm also very interested in this branch, as I am hoping it will improve performance on an ios-hosted map using apple's WKWebView. It renders google maps via the web very quickly, but seems to lag on tile rendering using v0.14.2 of mapbox-gl-js.  Mapbox of course puts gmaps to shame, so I *really* want to push the envelope there.  I'm not using the native mapbox because of missing features and plugins I want to leverage. 

Question for @averas -- how are you referencing this branch in your code? Is there a hosted version of the script or do I need to pull it down and serve it with my assets?

Also -- are there any features/fixes on master you wish you could use while on this branch?

For everyone else -- really looking forward to seeing this merged in!

**Comment by averas** [comment]
> how are you referencing this branch in your code? Is there a hosted version of the script or do I need to pull it down and serve it with my assets?

Yes, you will have to build and host it yourself. However, given the structure of the project, this is an easy thing to do.

> Also -- are there any features/fixes on master you wish you could use while on this branch?

My current routine is to rebase on <code>master</code>  in my own fork whenever regular releases are made. Whenever conflicts arise beyond my power to resolve I cry out here..

This routine keeps my own copy of <code>earcut-final</code> up to date with master so there are no features missing. There are, however, minor rendering issues (like antialiasing mentioned above) present. So far none of these issues are severe enough in my use cases though.

It should also be mentioned that this branch does not simply provide a performance boost in general but rather shines in cases where you use styles with a lot of polygon-fills. The Mapbox Streets style itself contains some of these (mostly the landcovers) so you will notice a difference there, but the real difference comes if you add on a lot of polygon-fills yourself.

**Comment by dchersey** [comment]
Awesome, thanks!  I’ll be using streets primarily … hopefully this gets the job done!

**Comment by dchersey** [comment]
Quick question — I’ve pulled down & built earcut-final, but I’m not seeing any directions on how to set up hosting. I have a node server in the cloud and I’m able to hit the debug page — how do I serve up the javascript for my app?

**Comment by mikemorris** [comment]
Went to rebase this onto `master` and hit a few conflicts - should i attempt rebasing here or do you want to update this @mourner?

**Comment by mourner** [comment]
I rebased it just a few days ago, wondering what could cause a lot of conflicts... If they look intimidating, I'll rebase tomorrow.

**Comment by mikemorris** [comment]
@mourner It might be from moving code into different files, but there's some GL clipping mask calls that look like they just get removed in `js/render/draw_fill.js` for example:
```
 73 <<<<<<< e7b551ccc8c36cf2031684398b7ae7b28fd9fa0d
 74     // Draw the stencil mask.
 75     painter.setDepthSublayer(1);
 76
 77     // We're only drawFilling to the first seven bits (== support a maximum of
 78     // 8 overlapping polygons in one place before we get rendering errors).
 79     gl.stencilMask(0x07);
 80     gl.clear(gl.STENCIL_BUFFER_BIT);
 81
 82     // Draw front facing triangles. Wherever the 0x80 bit is 1, we are
 83     // increasing the lower 7 bits by one if the triangle is a front-facing
 84     // triangle. This means that all visible polygons should be in CCW
 85     // orientation, while all holes (see below) are in CW orientation.
 86     painter.enableTileClippingMask(coord);
 87
 88     // When we do a nonzero fill, we count the number of times a pixel is
 89     // covered by a counterclockwise polygon, and subtract the number of
 90     // times it is "uncovered" by a clockwise polygon.
 91     gl.stencilOpSeparate(gl.FRONT, gl.KEEP, gl.KEEP, gl.INCR_WRAP);
 92     gl.stencilOpSeparate(gl.BACK, gl.KEEP, gl.KEEP, gl.DECR_WRAP);
 93
 94     // When drawFilling a shape, we first drawFill all shapes to the stencil buffer
 95     // and incrementing all areas where polygons are
 96     gl.colorMask(false, false, false, false);
 97     painter.depthMask(false);
 98
 99     // Draw the actual triangle fan into the stencil buffer.
100     gl.switchShader(painter.fillShader, translatedPosMatrix);
101
102     // Draw all buffers
103     var vertex = bucket.buffers.fillVertex;
104 =======
105     var vertex = tile.buffers.fillVertex;
106 >>>>>>> switch to tesselated fill rendering with earcut
```

**Comment by lucaswoj** [comment]
Fixes #2269

**Comment by mourner** [comment]
@lucaswoj @ansis can some of you rebase? I may not be the best person to resolve the new conflicts...

**Comment by ansis** [comment]
@mourner I'll try rebasing today

**Comment by ansis** [comment]
rebased 9369d51828d89d2ab0c7c30a792a7d91ce0afaa1 on master and then squashed into one commit.

Three render tests are failing:
- background-pattern/uneven-pattern: south america is covered by water
- extent/1024-fill: differences caused by test-suite bug https://github.com/mapbox/mapbox-gl-test-suite/issues/82
- fill-pattern/uneven-pattern: south america is covered by water

**Comment by mikemorris** [comment]
@ansis IIRC the `south america is covered by water` renders may be the result of using old V1 vector tiles in the test fixtures, [I saw this on gl-native too](https://github.com/mapbox/mapbox-gl-native/pull/2444#issuecomment-182974429).

**Comment by ansis** [comment]
I rebased earcut-final on master. Before the rebase it was a6c384af0df187de6045968f4539006d6212f434.

**Comment by jfirebaugh** [comment]
Next actions here are:

* Rebase on master once more
* Fix failing render tests
* Add a warning when encountering a v1 tile

File: js/source/worker_tile.js [hunk 1]
```diff
@@ -77,14 +77,15 @@ WorkerTile.prototype.parse = function(data, layerFamilies, actor, rawTileData, c
...
-            }
             layer = data.layers[sourceLayerId];
+            if (sourceLayerId === 'landcover') {
+                if (layer.version !== 2) {
+                    console.log({bad: '!!BAD!!', zoom: this.zoom, length: layer.length, coord: this.coord, version: layer.version});
+                    // console.warn('Vector tile source "' + this.source + '.' + sourceLayerId + '" does not use vector tile specification v2 and therefore may have some rendering errors.');
+                } else {
+                    console.log({bad: '@@GOOD@@', zoom: this.zoom, length: layer.length, coord: this.coord});
+                }
+            }
```
Comment (mourner) [inline js/source/worker_tile.js]: Did you commit this unintentionally?

Comment (lucaswoj) [inline js/source/worker_tile.js]: @mourner Intentionally (though perhaps not not in good practice) I was trying to debug some serverside problems with @springmeyer.

**Comment by lucaswoj** [comment]
This is ready to ship pending

 - [ ] a final code review of this PR 
 - [ ] a code review of https://github.com/mapbox/mapbox-gl-test-suite/pull/91

cc @mourner @jfirebaugh @jakepruitt @ansis @springmeyer

**Comment by jfirebaugh** [comment]
Let's eliminate `convertCoords`, rewriting `classifyRings` and the code that generates the `flattened` array for earcut in terms of the `{x: x, y: y}` point-geometry objects. This will both simplify the code and avoid GC churning a large number of 2-element arrays.

**Comment by mourner** [comment]
@jfirebaugh yep, done.

**Comment by mourner** [comment]
The test failed because of some really weird Git/NPM issue — tried [rebuilding without cache](https://circleci.com/gh/mapbox/mapbox-gl-js/3303) and it still failed with some "git bad object" errors. Not sure whether this is something temporary or broken in one of the subdeps.

**Comment by mourner** [comment]
Ah, it failed because the test suite ref it pointed to no longer exists for some reason.

**Comment by mourner** [comment]
OK, rebased again with updated test-suite and it passes.

**Comment by jfirebaugh** [comment]
👍 🚢 🎉

**Comment by springmeyer** [comment]
\o/ Thank you everyone!!

**Pull request merged**